### ISSUE TITLE
Question on creating a (for loop + if-else statement) combination

### ISSUE THREAD

**Issue by bobaekang** [uc-cfss/fp-bobaekang#1]
I am trying to create a column of binary variable/predictor for each Divvy station, proximity, which indicates whether a station is in proximity with any of CTA stops. I have two data sets, DivvyStations and CTAStops, both with the geographic coordinates (lon and lat) for each station/stop. I found that the `geosphere::distm()` function can be used to calculate the distance between two locations (in meters).

Using `distm()` to create the proximity classifier, I am trying to find a solution that combines for loop with if-else statement, but with little success.

The output should be a binary vector, where if a Divvy station is close to any CTA stop (for example, distance < 150 meters), 1, else, 0. The length of this vector should be equal to `nrow(DivvyStations)`.

The following is my trial,

> proximity = c()
> a = c()
> DivvyTest <- DivvyStation[1:10,]
> CTATest <- CTAStops[1:10,]
> for (i in DivvyTest) {
>   if (for (j in CTATest){
>     if(
>       150 < as.numeric(distm(c(DivvyStation[i,]$lon, DivvyStation[i,]$lat),
>             c(CTAStops[j,]$stop_lon, CTAStops[j,]$stop_lat),
>             fun = distHaversine))
>       ) {
>       a <- c(a, 1)
>       }
>     else {
>       a <- c(a, 0)
>       }
>   } && sum(a) != 0) {
>     proximity <- c(proximity, 1)
>     }
>   else {
>     proximity <- c(proximity, 0)
>     }
> }

Now, this loop gives me the following error:
> Error in .pointsToMatrix(x) : Wrong length for a vector, should be 2 

Please help me figure out how to fix this.
Thank you!

**Comment by bensoltoff** [comment]
My initial thought is that you're placing the conditional expressions in the wrong place. You really have a nested `for` loop with a conditional test _inside_ the second loop:

``` r
proximity = c()
DivvyTest <- DivvyStation[1:10,]
CTATest <- CTAStops[1:10,]
for (i in seq_along(DivvyTest)) {
  for (j in seq_along(CTATest)){
    if(distm(something[i], something[j]) < 150){
      proximity[[i]] <- TRUE
      break
    }
  }
}
```

Use the `break` command to exit the `j` loop as soon as a match is found. This way the first `TRUE` value doesn't get overwritten by another iteration where the distance isn't less than 150.

That pseudo-code should help you get started.

**Comment by bobaekang** [comment]
@bensoltoff Thank you very much. Based on your suggestion, I came up with the following:

``` r
proximity = c()
for (i in c(1:nrow(DivvyStation))) {
  for (j in c(1:nrow(CTAStops))){
    if(as.numeric(distm(c(DivvyStation[i,]$lon, DivvyStation[i,]$lat),
                        c(CTAStops[j,]$stop_lon, CTAStops[j,]$stop_lat),
                        fun = distHaversine)) < 150){
      proximity[[i]] <- 1
      break
    } else {
      proximity[[i]] <- 0
    }
  }
}
proximity
```

This seems to do the expected job though very slowly, probably due to the scale of this task (DivvyStation has 536 rows and CTAStops has 11520 rows).

**Comment by bensoltoff** [comment]
Yes because it has to loop through potentially 6,174,720 combinations.

It looks like the `distm` function creates a matrix of values. You could try running `distm` once without the loop.

``` r
divvy_m <- matrix(DivvyStation$lat, DivvyStation$lon)
cta_m <- matrix(CTAStops$lat, CTAStops$lon)

distances <- distm(divvy_m, cta_m, fun = distHaversine)
```

This should produce an 536 x 11520 matrix where each cell is the distance between the two stations designated for the row and column. Then just check each row to find if at least one cell is less than 150. This code should produce a vector of length 536 full of `TRUE/FALSE` values. Bind that to the `DivvyStations` data frame and you know which Divvy stations are near a CTA stop.

In general, R is much faster with matrix operations than looping over individual scalar values (like what you are doing now).

**Comment by bobaekang** [comment]
@bensoltoff  Thank you for your suggestion on using matrices instead. The following is what I came up with and it works much faster:

``` r
Divvy_m <- cbind(DivvyStation$lon, DivvyStation$lat)
CTA_m <- cbind(CTAStops$stop_lon, CTAStops$stop_lat)
distance_m <- distm(Divvy_m, CTA_m, fun = distHaversine)

proximity = c()
for (i in c(1:nrow(distance_m))) {
  for (j in c(1:ncol(distance_m))) {
    if (distance_m[i,j] < 150) {
      proximity[[i]] <- 1
      break
    } else {
      proximity[[i]] <- 0
    }
  }
}
```

**Comment by bensoltoff** [comment]
``` r
Divvy_m <- cbind(DivvyStation$lon, DivvyStation$lat)
CTA_m <- cbind(CTAStops$stop_lon, CTAStops$stop_lat)
distance_m <- distm(Divvy_m, CTA_m, fun = distHaversine)

# check if less than 150
distance_150 <- distance_m < 150

# divvy stations within 150 feet of CTA station
proximate <- rowSums(distance_150) > 0
```

I don't have your data, but I think this accomplishes the same task without any loops (and therefore even faster). Just turn `distance_m` into a logical matrix of `TRUE/FALSE`, then use `rowSums` to add up the number of `TRUE` values in each row. Another conditional check to see if this is greater than 0, and you have a vector that tells you if the Divvy station is near a CTA station.

**Comment by bobaekang** [comment]
bensoltoff Thank so much you for your help. I think this is a solution far superior to using any loop.  I wanted to keep the outcome a numeric vector of 1 or 0, so I did the following:

``` r
Divvy_m <- cbind(DivvyStation$lon, DivvyStation$lat)
CTA_m <- cbind(CTAStops$stop_lon, CTAStops$stop_lat)
distance_m <- distm(Divvy_m, CTA_m, fun = distHaversine) # a 535 by 11520 matrix for distance

distance150 <- distance_m <= 150 # check if the distance is <=150m
proximity150 <- (rowSums(distance150) > 0)*1 # a Divvy station is <=150m from any CTA stop, 1; otherwise, 0
```
